src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/FrameworkParametersCompletionProvider.cs

```diff
@@ -0,0 +1,536 @@
+    public ImmutableHashSet<char> TriggerCharacters { get; } = ImmutableHashSet.Create(' ');
+
+    public override bool ShouldTriggerCompletion(SourceText text, int caretPosition, CompletionTrigger trigger, OptionSet options)
+    {
+        if (trigger.Kind == CompletionTriggerKind.Insertion)
+        {
+            return TriggerCharacters.Contains(trigger.Character);
+        }
+
+        return false;
+    }
+
+    public override Task<CompletionDescription> GetDescriptionAsync(Document document, CompletionItem item, CancellationToken cancellationToken)
+    {
+        if (!item.Properties.TryGetValue(DescriptionKey, out var description))
+        {
+            return Task.FromResult<CompletionDescription>(null);
+        }
+
+        return Task.FromResult(CompletionDescription.Create(
+            ImmutableArray.Create(new TaggedText(TextTags.Text, description))));
+    }
+
+    public override Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey, CancellationToken cancellationToken)
+    {
+        // These values have always been added by us.
+        var startString = item.Properties[StartKey];
+        var lengthString = item.Properties[LengthKey];
+        var newText = item.Properties[NewTextKey];
+
+        // This value is optionally added in some cases and may not always be there.
+        item.Properties.TryGetValue(NewPositionKey, out var newPositionString);
+
+        return Task.FromResult(CompletionChange.Create(
+            new TextChange(new TextSpan(int.Parse(startString, CultureInfo.InvariantCulture), int.Parse(lengthString, CultureInfo.InvariantCulture)), newText),
+            newPositionString == null ? null : int.Parse(newPositionString, CultureInfo.InvariantCulture)));
+    }
+
+    public override async Task ProvideCompletionsAsync(CompletionContext context)
+    {
+        if (context.Trigger.Kind is not CompletionTriggerKind.Insertion)
+        {
+            return;
+        }
+
+        var position = context.Position;
+
+        var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+        if (root == null)
+        {
+            return;
+        }
+        var token = root.FindToken(position);
+
+        // If space is before a close paren then change current token to the previous argument.
+        if (token.IsKind(SyntaxKind.CloseParenToken) || token.IsKind(SyntaxKind.CommaToken))
+        {
+            token = token.GetPreviousToken();
+        }
+
+        // If space is after ? then it's likely a nullable type and move to previous type token.
+        if (token.IsKind(SyntaxKind.QuestionToken))
+        {
+            token = token.GetPreviousToken();
+        }
+
+        // Whitespace should follow the identifier token of the parameter.
+        if (!SyntaxFacts.IsPredefinedType(token.Kind()) &&
+            !token.IsKind(SyntaxKind.IdentifierToken))
+        {
+            return;
+        }
```

Comment:
do you care about generics?  e.g. `IList<int> $$`?  the prior token will not be a predefinedtype/identifier in that case.